File: android/src/main/java/com/aloha/kimage/KImagePlugin.java
package com.aloha.kimage;

import android.os.Environment;

import com.aloha.kimage.loader.AndroidUtilities;
import com.aloha.kimage.loader.ImageLoaderJava;
import com.aloha.kimage.loader.LoadImageCallback;

import java.io.File;
import java.lang.ref.WeakReference;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

public class KImagePlugin implements MethodChannel.MethodCallHandler {

    public String externalAbsolutePath() {
        return Environment.getExternalStorageDirectory().getAbsolutePath();
    }

    public static void registerWith(PluginRegistry.Registrar registrar) {
        AndroidUtilities.context = registrar.activeContext().getApplicationContext();
        MethodChannel channel = new MethodChannel(registrar.messenger(), "k_image");
        channel.setMethodCallHandler(new KImagePlugin());
    }

    @Override
    public void onMethodCall(final MethodCall call, final MethodChannel.Result result) {
        if (call.method.equals("documentsFolder")) {
            result.success(externalAbsolutePath());
            return;
        }

        String localPath = call.argument("path");
        int width = call.hasArgument("width") ? ((int)call.argument("width")) : -1;
        int height = call.hasArgument("height") ? ((int) call.argument("height")) : -1;

        if (localPath == null) {
            result.error("LocalPath is empty", null, null);
            return;
        }

        File file = new File(localPath);
        if (!file.exists()) {
            result.error("LocalPath is empty", null, null);
            return;
        }

        int type = -1;
        if (call.method.equals("fetchArtworkFromLocalPath")) {
            type = ImageLoaderJava.MEDIA_DIR_AUDIO;

        } else if (call.method.equals("loadImageFromLocalPath")) {
            type = ImageLoaderJava.MEDIA_DIR_IMAGE;
        } else if (call.method.equals("fetchVideoThumbnailFromLocalPath")) {
            type = ImageLoaderJava.MEDIA_DIR_VIDEO;
        }

        if (type == -1) {
            result.notImplemented();
            return;
        }

        int index = localPath.lastIndexOf('/');
        final String name = localPath.substring(index + 1);
        ImageLoaderJava.instance.generateWebThumbnail(type, localPath, name, width, height, new LoadImageCallback(name, result));
    }
}

File: android/src/main/kotlin/com/aloha/kimage/KImagePlugin.kt
package com.aloha.kimage

import android.os.Environment
import com.aloha.kimage.loader.AndroidUtilities
import com.aloha.kimage.loader.ImageLoader
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.File

class KImagePlugin : MethodCallHandler {
    private val externalAbsolutePath by lazy {
        Environment.getExternalStorageDirectory().absolutePath
    }

    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            AndroidUtilities.context = registrar.activeContext().applicationContext
            val channel = MethodChannel(registrar.messenger(), "k_image")
            channel.setMethodCallHandler(KImagePlugin())
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "documentsFolder") {
            result.success(externalAbsolutePath)
            return
        }

        val localPath = call.argument<String>("path")
        val width = call.argument<Int>("width")
        val height = call.argument<Int>("height")

        val file = File(localPath)
        if (localPath == null || !file.exists()) {
            result.error("LocalPath is empty", null, null)
            return
        }

        var type = -1
        if (call.method == "fetchArtworkFromLocalPath") {
            type = ImageLoader.MEDIA_DIR_AUDIO

        } else if (call.method == "loadImageFromLocalPath") {
            type = ImageLoader.MEDIA_DIR_IMAGE
        }

        if (type == -1) {
            result.notImplemented()
            return
        }

        GlobalScope.launch {
            ImageLoader.generateWebThumbnail(type, localPath, localPath.substringAfterLast("/"), width, height)?.let {
                result.success(it)
            }
        }
    }
}
